CONTRIBUTING document (#240)
ansible-collections/servicenow.itsm: CONTRIBUTING.md

# Contributing

## Getting Started

General information about setting up your Python environment, testing modules,
Ansible coding styles, and more can be found in the [Ansible Community Guide](
https://docs.ansible.com/ansible/latest/community/index.html).

## ServiceNow ITSM Collections

### servicenow.itsm

> The Ansible Collection for ServiceNow IT Service Management (ITSM) includes a variety of Ansible content to help automate the management of ServiceNow IT Service Management.

Reviewer: ```suggestion
The Ansible Collection for ServiceNow IT Service Management ([ITSM](https://www.servicenow.com/products/itsm.html)) includes a variety of Ansible content to help automate the management of ServiceNow IT Service Management.
```

New modules and plugins developed by the community should be proposed to `servicenow.itsm`.

## Submitting Issues
All software has bugs, and the `servicenow.itsm` collection is no exception. When you find a bug,
you can help tremendously by [telling us about it](https://github.com/ansible-collections/servicenow.itsm/issues).

If you should discover that the bug you're trying to file already exists in an issue,
you can help by verifying the behavior of the reported bug with a comment in that
issue, or by reporting any additional information.

## Writing New Code

> New modules should be submitted to the [servicenow.itsm](https://github.com/ansible-collections/servicenow.itsm) collection.

Reviewer 2: This collection is more than modules. I'd prefer the use of "plugins" or "modules and plugins" here and throughout this doc. Other collections make that mistake and I am personally trying to stamp that out.

> ### Static analysis

Reviewer 2: This section title doesn't seem right to me. I thought this doc didn't address new features and bug fixes until I looked again. kubernetes.core just has a ["Pull Requests" section](https://github.com/ansible-collections/kubernetes.core/blob/main/CONTRIBUTING.md#pull-requests) the encompasses new code and features/bugs. amazon.aws has it all under one ["Writing new code" section](https://github.com/ansible-collections/kubernetes.core/blob/main/CONTRIBUTING.md#pull-requests).

> For new features and bug fixes on existing modules, clone this repository and try to run unit tests and integration tests by following [these instructions](https://docs.ansible.com/ansible/latest/community/create_pr_quick_start.html). When you get to this part, we use a wide variety of linters and static code analysis tools to ensure our code base stays healthy. To run all checks, we can execute the following command:

Reviewer 2: "plugins" or "modules and plugins" again.

```
(venv) $ make sanity
```
This command will take care of installing requirements and running them.

### Unit tests

> We use unit tests to test the logic and implementation of our helper functions and various Ansible plugins that are part of the Sensu Go collection.

Reviewer: ```suggestion
We use unit tests to test the logic and implementation of our helper functions and various Ansible plugins that are part of the `servicenow.itsm` collection.
```

```
(venv) $ make units
```

> ### Integration tests

Reviewer 2: I think there needs to be something about ServiceNow and setting up their test environment for this? Not sure that is the case and if so that this is the right place for it. Thinking as a developer that is a question I have after reading this -- what about my ServiceNow setup to test?

Also should this doc mention anything about the ServiceNow app or at least link to something regarding that?

Reviewer 3: For me, integration tests against free developer servicenow instance were super slow, especially when it arrived to configuration modules. The connection got interrupted from time to time for various reasons including SNOW instance hibernation in the middle of the test. It would be nice to have some tips how to make it work as good as possible.

The link to SNOW developer page may be also fine but I think that more details are not necessary. There are guides on SNOW pages how to set up SNOW testing environment. If this somehow enters documentation here, the changes on the SNOW side could outdate information here. I think reference to [General Usage Patterns](https://github.com/ansible-collections/servicenow.itsm/blob/main/docs/general_usage_patterns.md) is quite enough.

Author: I was planning on writing a quick guide on how you can Set up your ServiceNow instance and OAuth credentials in the CONTRIBUTING doc, and then reference to [General Usage Patterns](https://github.com/ansible-collections/servicenow.itsm/blob/main/docs/general_usage_patterns.md) for more information.

* Resource deletion under check mode
* Resource deletion
* Resource deletion (of a non-existent resource) under check mode
* Resource deletion (of a non-existent resource)

Where modules have multiple parameters we recommend running through the 4-step modification cycle for each parameter the module accepts, as well as a modification cycle where as most, if not all, parameters are modified at the same time.

For general information on running the integration tests see the
[Integration Tests page of the Module Development Guide](https://docs.ansible.com/ansible/devel/dev_guide/testing_integration.html#testing-integration). For questions about writing tests the Ansible ServiceNow ITSM can be found on Libera.Chat IRC as detailed below.

### Code of Conduct
The `servicenow.itsm` collection follows the Ansible project's
[Code of Conduct](https://docs.ansible.com/ansible/devel/community/code_of_conduct.html).
Please read and familiarize yourself with this document.

### IRC
Our IRC channels may require you to register your nickname. If you receive an error when you connect, see
[Libera.Chat's Nickname Registration guide](https://libera.chat/guides/registration) for instructions.

> The `#ansible-community` channel on [libera.chat](https://libera.chat/) IRC is the main and official place to discuss use and development of the `ServiceNow.ITSM` collection.

Reviewer: ```suggestion
The `#ansible-community` channel on [libera.chat](https://libera.chat/) IRC is the main and official place to discuss the use and development of the `serviceNow.itsm` collection.
```

you can help tremendously by [telling us about it](https://github.com/ansible-collections/servicenow.itsm/issues).

If you should discover that the bug you're trying to file already exists in an issue,
you can help by verifying the behavior of the reported bug with a comment in that
issue, or by reporting any additional information.

## Environment setup
1. Sign up for a [ServiceNow ID](https://signon.service-now.com/x_snc_sso_auth.do#/sign-up) 
1. Request and setup an Instance. 
1. When the instance is ready, you will be redirected to the Developer Portal `e.g. https://dev12345.service-now.com/`. In All, search for Application Registry then New - Create an OAuth API endpoint for external clients.
1. Ensure that you have changed the Integration tests configuration file accordingly with your Personal Developer Instance credentials for connection and access. A template is available [here](https://github.com/ansible-collections/servicenow.itsm/blob/main/tests/integration/integration_config.yml.template).
1. For more information, check [General usage pattenrs](https://github.com/ansible-collections/servicenow.itsm/blob/main/docs/general_usage_patterns.md).

> 

Reviewer 3: I think it is necessary to mention here that docker or podman is installed locally as all tests (sanity/units/integration) are executed in docker when ran through 'make'.

Reviewer 2: That sounds reasonable.
